qip-0014.md

```diff
@@ -0,0 +1,170 @@
+```
+ QIP: 14
+ Layer: EVM
+ Title: CREATE + CREATE2 opcode
+ Author: alanorwick <alan+qip@quai.org>
+ Comments-Summary: No comments yet.
+ Comments-URI: https://github.com/quainetwork/qips/wiki/Comments:QIP-0014
+ Status: Draft
+ Type: Informational
+ Created: 2024-07-31
+ License: BSD-2-Clause
+```
+
+
+## Abstract
+
+This QIP proposes modifications to the CREATE opcode in Quai Network to ensure shard-specific address generation, while maintaining CREATE2 functionality as is. The proposal aims to adapt Ethereum's address creation mechanism to Quai's sharded architecture.
```

Comment:
This is a great point. I would query @jdowning100 or @mechanikalk on the downstream impact on contract deployment. Many contracts that use create (AMMS/NFTs/proxy contracts/etc.) are using solidity and expecting `CREATE` to work. We would likely need to add this new `CREATE3` opcode in solidity as well then.

Reply:
The middle ground which I proposed to @mechanikalk is hashing all of the bytecode during the grinding since then the `CREATE` computation `keccak256(([sender_address, sender_nonce, bytecode])[12:]` fits the prior spec. If we include the grinding, this comes at the cost of the keccak on the entire length of the bytecode which is more expensive.

Reply:
Yeah, I understand this would require updating any contract which uses CREATE to use CREATE3 instead, but that is definitely the preferred way to do this. We can even deprecate CREATE and CREATE2 if we think that those methods of contract deployment are ill-advised.

Deprecating old opcodes and supplanting them with new opcodes is the way to extend an instruction set, like the EVM. Changing the behavior of existing opcodes is dangerous.

Reply:
I invite @mechanikalk @jdowning100 to weigh in

Reply:
I agree on @mechanikalk and @jdowning100 weighing in. We implemented the current CREATE logic during Iron Age in order to get AMMs / contracts deploying contracts to work in a minimal fashion. Relying on Chesterton's Fence principle we should further assess what caused us to implement it in CREATE versus making CREATE3.